plugin/http/httpstats/transport.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpstats
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+)
+
+// Transport is an http.RoundTripper that collects stats for the outgoing requests.
+type Transport struct {
+	// Base represents the underlying roundtripper that does the actual requests.
+	// If none is given, http.DefaultTransport is used.
+	//
+	// If base HTTP roundtripper implements CancelRequest,
+	// the returned round tripper will be cancelable.
+	Base http.RoundTripper
+}
+
+// RoundTrip implements http.RoundTripper, delegating to Base and recording stats for the request.
+func (t Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx, _ := tag.New(req.Context(),
+		tag.Upsert(Host, req.URL.Host),
+		tag.Upsert(Path, req.URL.Path),
+		tag.Upsert(Method, req.Method))
+	req = req.WithContext(ctx)
+	var (
+		resp *http.Response
+		err  error
+	)
+	track := &tracker{
+		start: time.Now(),
+		ctx:   ctx,
+	}
+	if req.Body == nil {
+		//TODO: handle cases where ContentLength is not set
+		track.reqSize = -1
+	} else if req.ContentLength > 0 {
+		track.reqSize = req.ContentLength
+	}
+	stats.Record(ctx, ClientRequest.M(1))
+
+	// Perform request
+	resp, err = t.base().RoundTrip(req)
+
+	if err != nil {
+		track.statusCode = "error"
+		track.end()
+	} else {
+		track.statusCode = strconv.Itoa(resp.StatusCode)
+		if resp.Body == nil {
+			track.end()
+		} else {
+			track.body, resp.Body = resp.Body, track
```

Comment:
We could make this clearer perhaps by
```go
track.body = resp.Body
resp.Body = track
```
I mention this because I had to scan the entire statement and read it twice.

Reply:
ack